ibis/util.py

```diff
@@ -457,106 +458,90 @@ def wrapper(*args, **kwargs):
+        return len(self._data)
 
-    @abc.abstractmethod
-    def __component_eq__(
-        self,
-        other: Any,
-        cache: MutableMapping[Hashable, bool],
-    ) -> bool:
-        """Compare the individual fields of a subclass."""
+    def __iter__(self):
+        return iter(self._data)
 
+    def __setitem__(self, key, value):
+        # construct an alternative representation of the key using the id()
+        # of the key's components, this prevents infinite recursions
+        identifiers = tuple(id(item) for item in key)
 
-C = TypeVar("C", bound=EqMixin)
+        # create a function which removes the key from the cache
+        def callback(ref_):
+            return self._data.pop(identifiers, None)
 
+        # create weak references for the key's components with the callback
+        # to remove the cache antry if any of the key's components gets
+        # garbase collected
+        refs = tuple(weakref.ref(item, callback) for item in key)
 
-def _try_eq(
-    left: Any,
-    right: Any,
-    *,
-    cache: MutableMapping[Hashable, bool],
-) -> bool:
-    try:
-        return (left is None and right is None) or left.equals(
-            right,
-            cache=cache,
-        )
-    except AttributeError:
-        return left == right
+        self._data[identifiers] = (value, refs)
 
+    def __getitem__(self, key):
+        identifiers = tuple(id(item) for item in key)
+        value, _ = self._data[identifiers]
+        return value
 
-def seq_eq(
-    lefts: Sequence[C],
-    rights: Sequence[C],
-    *,
-    cache: MutableMapping[Hashable, bool],
-) -> bool:
-    """Compare two sequences of expressions.
+    def __delitem__(self, key):
+        identifiers = tuple(id(item) for item in key)
+        del self._data[identifiers]
 
-    Parameters
-    ----------
-    lefts
-        Iterable of expressions
-    rights
-        Iterable of expressions
-    cache
-        Mutable mapping of expression pairs to bool
+    def __repr__(self):
+        return repr(self._data)
 
-    Returns
-    -------
-    bool
-        Whether the expressions of `lefts` and `rights` are equal
-    """
-    return len(lefts) == len(rights) and all(
-        _try_eq(left, right, cache=cache) for left, right in zip(lefts, rights)
-    )
+
+class Comparable(abc.ABC):
+
+    __slots__ = ("__weakref__",)
+    __cache__ = WeakCache()
+
+    def __hash__(self):
+        return super().__hash__()
+
+    def __eq__(self, other):
+        try:
+            return self.__cached_equal__(other)
+        except TypeError:
+            raise NotImplemented  # noqa: F901
+
+    @abc.abstractmethod
+    def __equal__(self, other):
+        ...
+
+    def __cached_equal__(self, other):
+        if self is other:
+            return True
+
+        # type comparison should be cheap
+        if type(self) != type(other):
+            return False
+
+        # reduce space required for commutative operation
+        if hash(self) < hash(other):
```

Comment:
Since equality is commutative the order of the arguments doesn't matter, so this is an optimisation to store only `(a, b)` instead of both `(a, b)` and `(b, a)`